### ISSUE TITLE
CMake v3.20.0 causes build to fail

### ISSUE THREAD

**Issue by zx8** [fish-shell/fish-shell#7862]
```
$ cmake . \
  -DCMAKE_C_FLAGS_RELEASE=-DNDEBUG \
  -DCMAKE_CXX_FLAGS_RELEASE=-DNDEBUG \
  -DCMAKE_INSTALL_PREFIX=/usr/local/Cellar/fish/HEAD-cd1f0cc \
  -DCMAKE_INSTALL_LIBDIR=lib \
  -DCMAKE_BUILD_TYPE=Release \
  -DCMAKE_FIND_FRAMEWORK=LAST \
  -DCMAKE_VERBOSE_MAKEFILE=ON \
  -Wno-dev \
  -DCMAKE_OSX_SYSROOT=/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk \
  -Dextra_functionsdir=/usr/local/share/fish/vendor_functions.d \
  -Dextra_completionsdir=/usr/local/share/fish/vendor_completions.d \
  -Dextra_confdir=/usr/local/share/fish/vendor_conf.d \
  -DSED=/usr/bin/sed

$ make install
<snip>
/usr/local/Cellar/cmake/3.20.0/bin/cmake -E cmake_link_script CMakeFiles/fish_indent.dir/link.txt --verbose=1
/usr/local/Homebrew/Library/Homebrew/shims/mac/super/clang++  -Wall -Wextra -Wno-comment -Wno-address  -Wredundant-move -isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk -mmacosx-version-min=10.9 -Wl,-search_paths_first -Wl,-headerpad_max_install_names CMakeFiles/fish_indent.dir/src/fish_indent.cpp.o CMakeFiles/fish_indent.dir/src/print_help.cpp.o -o fish_indent  libfishlib.a /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/lib/libcurses.tbd /usr/local/lib/libpcre2-32.dylib
Undefined symbols for architecture x86_64:
  "_libintl_bindtextdomain", referenced from:
      fish_bindtextdomain(char const*, char const*) in libfishlib.a(fallback.cpp.o)
  "_libintl_gettext", referenced from:
      fish_gettext(char const*) in libfishlib.a(fallback.cpp.o)
  "_libintl_textdomain", referenced from:
      fish_textdomain(char const*) in libfishlib.a(fallback.cpp.o)
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[3]: *** [fish_indent] Error 1
make[2]: *** [CMakeFiles/fish_indent.dir/all] Error 2
make[1]: *** [all] Error 2
make: *** [install] Error 2
```

**Comment by zanchey** [comment]
Yes, I see this on macOS 10.14 with cd1f0cc5d when building under Homebrew (please include these details!), but not under a standalone build. Do you have a `/Library/Frameworks/Mono.framework/` directory? If so, this is probably a duplicate of #5244.

**Comment by zx8** [comment]
I do have a `/Library/Frameworks/Mono.framework/` directory, and you're correct, this is a Homebrew build (sorry, raised this in a rush before heading to work).

The odd thing my last successful build was:

```
$ fish --version
fish, version 3.1.2-2282-gbcbfd70d4
```

So unless something has changed in Homebrew or perhaps my local environment in that time, I can't see why this would have started failing.

**Comment by zanchey** [comment]
Homebrew and CMake have both had releases recently, so it's certainly possible. CMake 3.20 has some updates to the `FindIntl` module, which I suspect might be related.

**Comment by zx8** [comment]
Tried the solution suggested @ https://github.com/EOSIO/eos/issues/2028#issuecomment-378258329 but no luck.

**Comment by zx8** [comment]
Tried the solution suggested @ https://github.com/EOSIO/eos/issues/2174#issuecomment-379210407 but no luck.

**Comment by zx8** [comment]
Reverting CMake back to v3.19.7 resolves the issue, so it definitely seems to be a bug with v3.20.0.

**Comment by zanchey** [comment]
Yes, this is the culprit:

https://github.com/Kitware/CMake/commit/c30d06b7e66c9bfe52f91f472a38cefdfc527601

The problem is that `check_symbol_exists` always [returns true if the symbol is defined as a macro](https://cmake.org/cmake/help/latest/module/CheckSymbolExists.html) without checking if it needs to be linked. So `Intl_IS_BUILTIN` gets set without checking the linking thanks to the `_INTL_REDIRECT_MACROS` definition in `libintl.h`.

**Comment by zanchey** [comment]
Also excitingly, `libintl.h` is different under C or C++ compilers, and the `FindPackage(Intl)` call uses C symbols.

**Comment by zanchey** [comment]
Filed upstream at https://gitlab.kitware.com/cmake/cmake/-/issues/21979

**Comment by mqudsi** [comment]
Awesome catch, @zanchey. It’s been fixed for 3.20.1
https://gitlab.kitware.com/cmake/cmake/-/merge_requests/5948

**Comment by faho** [comment]
Okay, unfortunately there is nothing more we can do (apart from @zanchey's awesome bug reporting), so I'm closing this. Hopefully cmake 3.20.1 will be out soon and every distribution will actually upgrade.
